File: spikes/828-passive-observer/prototype/extract.py
"""SessionFacts (with its child tree) -> RunFacts (spike #828, CONTRACTS.md "extract.py").

Harness-neutral derivation of per-run facts. Everything here is arithmetic over the typed facts a
source produced; no session content exists at this layer to leak.

Scope of "the run" (the contract leaves this implicit, so it is stated here):
  - Counts, token totals, invocations, in-band assets and forbids merge over the WHOLE tree
    (main transcript plus every linked child transcript). Usages are deduplicated by provider
    message id across the tree so a response logged in two places counts once.
  - `run_outcome`, `turns`, `loaded_events` and `truncated` describe the MAIN transcript only: a
    child's "user" lines are the parent's task prompt, not a person's turns, and the loaded set the
    segments are cut from is the parent's.
  - A parent `Agent` call is linked to a direct child by `child_key` (toolUseResult.agentId) or by
    the child's `child_meta["toolUseId"]` (D4). Its InvocationObs carries the child's exact token
    total and the child's outcome; a linked child's own tokens are ALSO in the run totals.
  - "Total tokens" of a child = input + output + cache_creation + cache_read. `cached_input`,
    `thinking` and `reasoning` are subsets of input/output for the providers that report them and
    are excluded so nothing is counted twice.
"""
from __future__ import annotations

from collections import Counter
from datetime import datetime, timezone
from typing import Dict, Iterator, List, Optional

import taskcat
from model import ASSET_AGENT, ASSET_MCP_SERVER, ASSET_SKILL, InvocationObs, RunFacts
from sources.base import (
    FAILURE_INTERRUPTED,
    FAILURE_UNKNOWN,
    FAILURE_USER_DENIED,
    RATE_BEARING_FAILURES,
    SessionFacts,
    ToolCall,
    Usage,
)

OUTCOME_TRUNCATED = "truncated"
OUTCOME_COMPACTED = "compacted"
OUTCOME_INTERRUPTED = "interrupted"
OUTCOME_COMPLETED = "completed"
OUTCOME_UNKNOWN = "unknown"

TOOL_CLASSES = ("edit", "read", "shell", "mcp", "other")
EDIT_TOOLS = frozenset({"Edit", "Write", "MultiEdit", "NotebookEdit", "apply_patch"})
READ_TOOLS = frozenset({"Read", "Glob", "Grep", "LS", "WebFetch", "WebSearch"})
SHELL_TOOLS = frozenset({"Bash", "shell", "exec"})
REPEAT_THRESHOLD = 3

PERMISSION_MODES = {
    "acceptEdits": "accept_edits",
    "bypassPermissions": "bypass",
    "dontAsk": "dont_ask",
    "plan": "plan",
    "default": "default",
    "auto": "auto",
}
EFFORTS = frozenset({"minimal", "low", "medium", "high", "xhigh"})

# (envelope key, Usage attribute); the first two are never null on the wire.
TOKEN_BUCKETS = (
    ("input", "input_tokens"),
    ("output", "output_tokens"),
    ("cache_creation", "cache_creation"),
    ("cache_read", "cache_read"),
    ("cached_input", "cached_input"),
    ("thinking", "thinking"),
    ("reasoning", "reasoning"),
)
NON_NULL_BUCKETS = ("input", "output")
TOTAL_BUCKETS = ("input", "output", "cache_creation", "cache_read")


def extract(facts: SessionFacts, now_ms: int) -> RunFacts:
    """Derive RunFacts for the run rooted at `facts`. `now_ms` is only the fallback for a session
    that carries no harness timestamp at all."""
    tree = list(walk(facts))
    calls = [c for f in tree for c in f.tool_calls]
    usages = dedupe_usages(tree)
    first, last = _span(tree, now_ms)
    return RunFacts(
        session_key=facts.ref.session_key,
        harness=facts.ref.harness,
        harness_version=facts.harness_version,
        entrypoint_class=entrypoint_class(facts.entrypoint),
        effort=effort_class(facts.effort),
        permission_mode=permission_mode(facts.permission_mode),
        model=taskcat.allowlist_model(dominant_model(tree)),
        observed_day=utc_day(first),
        first_ts_ms=first,
        last_ts_ms=last,
        run_outcome=run_outcome(facts),
        turns=facts.user_turns,
        tool_calls=len(calls),
        tool_failures=sum(1 for c in calls if c.failure_class in RATE_BEARING_FAILURES),
        user_denials=sum(1 for c in calls if c.failure_class == FAILURE_USER_DENIED),
        subagent_runs=len(tree) - 1,
        compactions=sum(f.compactions for f in tree),
        unpaired_tool_uses=sum(1 for c in calls if not c.paired),
        repeated_tool_calls=repeated_tool_calls(calls),
        tokens=sum_tokens(usages),
        tokens_by_model=sum_tokens_by_model(usages),
        mcp_corroborations=merge_mcp_corroborations(tree),
        tokens_basis="harness_usage" if usages else "none",
        tool_class_shares=tool_class_shares(calls),
        invocations=invocations(facts),
        loaded_events=list(facts.loaded_events),
        in_band_assets=[a for f in tree for a in f.in_band_assets],
        lines_seen=sum(f.lines_seen for f in tree),
        lines_unknown_type=sum(f.lines_unknown_type for f in tree),
        bytes_read=sum(f.bytes_read for f in tree),
        parse_errors=sum(f.parse_errors for f in tree),
        truncated=facts.truncated,
        forbids=merge_forbids(tree),
    )


# -- tree helpers --------------------------------------------------------------------------------


def walk(facts: SessionFacts) -> Iterator[SessionFacts]:
    """Depth-first, parent before children, in the order the source linked them."""
    yield facts
    for child in facts.children:
        yield from walk(child)


def dedupe_usages(tree: List[SessionFacts]) -> Dict[str, Usage]:
    """One Usage per provider message id across the tree. A streamed response is written as several
    lines whose usage grows as output streams, so the entry with the largest output count wins;
    on a tie the first occurrence (parent first) is kept."""
    seen: Dict[str, Usage] = {}
    for f in tree:
        for mid, usage in f.usages.items():
            current = seen.get(mid)
            if current is None or usage.output_tokens > current.output_tokens:
                seen[mid] = usage
    return seen


def merge_mcp_corroborations(tree: List[SessionFacts]) -> Dict[str, int]:
    out: Dict[str, int] = {}
    for f in tree:
        for server, n in f.mcp_attribution_counts.items():
            out[server] = out.get(server, 0) + n
    return out


def merge_forbids(tree: List[SessionFacts]) -> Dict[str, set]:
    out: Dict[str, set] = {}
    for f in tree:
        for bucket, values in f.forbids.items():
            out.setdefault(bucket, set()).update(values)
    return out


def _span(tree: List[SessionFacts], now_ms: int) -> "tuple[int, int]":
    firsts = [f.first_ts_ms for f in tree if f.first_ts_ms is not None]
    lasts = [f.last_ts_ms for f in tree if f.last_ts_ms is not None]
    first = min(firsts) if firsts else now_ms
    last = max(lasts) if lasts else first
    return first, max(first, last)


def dominant_model(tree: List[SessionFacts]) -> str:
    """Most frequent model by response count in the MAIN transcript (tree[0]); the whole tree is
    used only when the main transcript carried no model. Sub-agents may run on a different model,
    which is why the envelope also carries tokens_by_model. Ties break on the smaller name so the
    result is deterministic. "unknown" when no response carried a model."""
    counts: Counter = Counter()
    if tree and tree[0].models:
        counts.update(tree[0].models)
    else:
        for f in tree:
            counts.update(f.models)
    if not counts:
        return "unknown"
    return sorted(counts.items(), key=lambda kv: (-kv[1], kv[0]))[0][0]


# -- tokens --------------------------------------------------------------------------------------


def sum_tokens(usages: Dict[str, Usage]) -> Dict[str, Optional[int]]:
    """Envelope-shaped token totals. Nullable buckets stay None unless at least one usage reported
    a value for them, so a provider without that bucket is "absent", not zero."""
    out: Dict[str, Optional[int]] = {key: (0 if key in NON_NULL_BUCKETS else None) for key, _ in TOKEN_BUCKETS}
    for usage in usages.values():
        for key, attr in TOKEN_BUCKETS:
            value = getattr(usage, attr)
            if value is None:
                continue
            out[key] = (out[key] or 0) + value
    return out


def sum_tokens_by_model(usages: Dict[str, Usage]) -> Dict[str, Dict[str, Optional[int]]]:
    """Envelope-shaped token totals per allowlisted model id (sub-agents may run on another model)."""
    by_model: Dict[str, Dict[str, Usage]] = {}
    for mid, usage in usages.items():
        by_model.setdefault(taskcat.allowlist_model(usage.model), {})[mid] = usage
    return {model: sum_tokens(group) for model, group in sorted(by_model.items())}


def total_tokens(tokens: Dict[str, Optional[int]]) -> int:
    """input + output + cache_creation + cache_read (see module docstring for the exclusions)."""
    return sum(tokens.get(key) or 0 for key in TOTAL_BUCKETS)


def child_tokens_total(child: SessionFacts) -> Optional[int]:
    """Exact total of a child run from its own transcript tree; None when the child carries no
    usage record at all (no evidence is not zero)."""
    usages = dedupe_usages(list(walk(child)))
    if not usages:
        return None
    return total_tokens(sum_tokens(usages))


# -- tool calls ----------------------------------------------------------------------------------


def repeated_tool_calls(calls: List[ToolCall]) -> int:
    """Number of calls belonging to a (name, input_fingerprint) group of size >= REPEAT_THRESHOLD."""
    groups = Counter((c.name, c.input_fingerprint) for c in calls)
    return sum(n for n in groups.values() if n >= REPEAT_THRESHOLD)


def tool_class(name: str, server: Optional[str]) -> str:
    """Published tool-mix classification (D2). MCP is decided first: a Codex MCP tool is named
    `<server>__<tool>` without the `mcp__` prefix but has `server` set."""
    if server or name.startswith("mcp__"):
        return "mcp"
    if name in EDIT_TOOLS:
        return "edit"
    if name in READ_TOOLS:
        return "read"
    if name in SHELL_TOOLS:
        return "shell"
    return "other"


def tool_class_shares(calls: List[ToolCall]) -> Dict[str, float]:
    """count/total per class; every class present (0.0 when absent) so the shape is stable.
    All zeros when there are no calls, which taskcat maps to `unspecified`."""
    counts = Counter(tool_class(c.name, c.server) for c in calls)
    total = len(calls)
    return {cls: (counts[cls] / total if total else 0.0) for cls in TOOL_CLASSES}


def _latency(call: ToolCall) -> Optional[int]:
    return None if (call.is_async or not call.paired) else call.latency_ms


# -- run shape -----------------------------------------------------------------------------------


def entrypoint_class(raw: Optional[str]) -> str:
    s = (raw or "").lower()
    if "remote" in s:
        return "remote"
    if "vscode" in s or "jetbrains" in s or "ide" in s:
        return "ide"
    if "sdk" in s:
        return "sdk"
    if "cli" in s:
        return "cli"
    return "unknown"


PERMISSION_ENUM = ("default", "plan", "accept_edits", "bypass", "auto", "dont_ask", "unknown")


def permission_mode(raw: Optional[str]) -> str:
    """Claude Code raw values are mapped; a value already in the closed enum (a source that
    pre-maps, like the Codex approval policy) passes through unchanged."""
    if raw in PERMISSION_ENUM:
        return raw
    return PERMISSION_MODES.get(raw or "", "unknown")


def effort_class(raw: Optional[str]) -> str:
    """Closed enum from the gate; anything else (including harness values the gate does not list)
    is "unknown". The contract is silent on effort; this keeps the payload gate-clean."""
    return raw if raw in EFFORTS else "unknown"


def utc_day(ts_ms: int) -> str:
    return datetime.fromtimestamp(ts_ms // 1000, tz=timezone.utc).strftime("%Y-%m-%d")


def run_outcome(facts: SessionFacts) -> str:
    """Decision table, first match wins: truncated > compacted > interrupted > completed > unknown.
    Only the transcript `facts` itself is consulted (children have their own outcome)."""
    if facts.truncated:
        return OUTCOME_TRUNCATED
    if facts.compactions > 0 and facts.last_stop_reason != "end_turn":
        return OUTCOME_COMPACTED
    if _interrupted_at_end(facts.tool_calls):
        return OUTCOME_INTERRUPTED
    if facts.last_stop_reason == "end_turn":
        return OUTCOME_COMPLETED
    return OUTCOME_UNKNOWN


def _interrupted_at_end(calls: List[ToolCall]) -> bool:
    """Any call without a result, or the last call in transcript order marked interrupted."""
    if any(not c.paired for c in calls):
        return True
    return bool(calls) and calls[-1].interrupted


# -- invocations ---------------------------------------------------------------------------------


def invocations(facts: SessionFacts) -> List[InvocationObs]:
    """Explicit asset invocations over the tree: parent's calls first, then each child's."""
    out: List[InvocationObs] = []
    for call in facts.tool_calls:
        if call.skill:
            out.append(InvocationObs(asset_type=ASSET_SKILL, name=call.skill, ts_ms=call.ts_ms,
                                     latency_ms=_latency(call), failure_class=call.failure_class,
                                     is_async=call.is_async))
        elif call.server:
            out.append(InvocationObs(asset_type=ASSET_MCP_SERVER, name=call.server, ts_ms=call.ts_ms,
                                     latency_ms=_latency(call), failure_class=call.failure_class,
                                     is_async=call.is_async))
        elif call.agent_type:
            out.append(_agent_invocation(call, _linked_child(facts, call)))
    for child in facts.children:
        out.extend(invocations(child))
    return out


def _linked_child(facts: SessionFacts, call: ToolCall) -> Optional[SessionFacts]:
    for child in facts.children:
        if call.child_key and child.ref.session_key == call.child_key:
            return child
        if child.ref.child_meta.get("toolUseId") == call.tool_use_id:
            return child
    return None


def _agent_invocation(call: ToolCall, child: Optional[SessionFacts]) -> InvocationObs:
    """The parent's result is only a spawn ack (D4): outcome and tokens come from the child when
    one is linked. The parent's own failure class (a denied or failed spawn) still takes precedence
    because then there is no child run to speak of."""
    failure = call.failure_class
    corroborated = False
    tokens: Optional[int] = None
    if child is not None:
        failure = failure or _child_failure_class(child)
        corroborated = str(child.ref.child_meta.get("corroborated", "")).lower() == "true"
        tokens = child_tokens_total(child)
    return InvocationObs(asset_type=ASSET_AGENT, name=call.agent_type or "", ts_ms=call.ts_ms,
                         latency_ms=_latency(call), failure_class=failure, is_async=call.is_async,
                         corroborated=corroborated, child_tokens_total=tokens)


def _child_failure_class(child: SessionFacts) -> Optional[str]:
    """Child outcome -> failure class. Never a rate-bearing class: a child's own tool errors are
    counted on the tools it called, not on the agent as a whole."""
    outcome = run_outcome(child)
    if outcome == OUTCOME_INTERRUPTED:
        return FAILURE_INTERRUPTED
    if outcome in (OUTCOME_TRUNCATED, OUTCOME_UNKNOWN):
        return FAILURE_UNKNOWN
    return None

File: crates/vettd-cli/src/observe/extract.rs
//! [`SessionFacts`] (with its child tree) -> [`RunFacts`].
//!
//! Port of `spikes/828-passive-observer/prototype/extract.py`, function for function.
//!
//! Harness-neutral derivation of per-run facts. Everything here is arithmetic over the typed facts
//! a source produced; no session content exists at this layer to leak. The one field that could be
//! mistaken for content — [`RunFacts::tool_class_shares`] — is local only: it is the input to
//! [`taskcat::categorize`] and only the resulting closed category egresses.
//!
//! Scope of "the run" (the contract leaves this implicit, so `extract.py:6-18` states it and so
//! does this):
//!
//! - Counts, token totals, invocations, in-band assets and forbids merge over the WHOLE tree (main
//!   transcript plus every linked child transcript). Usages are deduplicated by provider message id
//!   across the tree so a response logged in two places counts once.
//! - [`run_outcome`], `turns`, `loaded_events` and `truncated` describe the MAIN transcript only: a
//!   child's "user" lines are the parent's task prompt, not a person's turns, and the loaded set the
//!   segments are cut from is the parent's.
//! - A parent `Agent` call is linked to a direct child by `child_key` (`toolUseResult.agentId`) or
//!   by the child's `child_meta["toolUseId"]` (D4). Its [`InvocationObs`] carries the child's exact
//!   token total and the child's outcome; a linked child's own tokens are ALSO in the run totals.
//! - "Total tokens" of a child = input + output + cache_creation + cache_read. `cached_input`,
//!   `thinking` and `reasoning` are subsets of input/output for the providers that report them and
//!   are excluded so nothing is counted twice.
//!
//! The token and tool-call arithmetic lives in the sibling `extract_tally.rs` and is re-exported
//! here, so callers still say `observe::extract::sum_tokens`: the split is the 400-line file budget
//! (CONTRIBUTING.md "File size limits") and carries no meaning of its own.
//!
//! **Named divergence from the Python.** Python integers are unbounded; these sums are `i64`/`u64`
//! and saturate. Every input is a token count, a byte length or a line count from one machine's
//! logs, so the saturation point is unreachable in practice; saturating rather than wrapping is
//! chosen so that an absurd transcript degrades a number instead of panicking a read.

#[path = "extract_tally.rs"]
mod tally;

use std::collections::{BTreeMap, BTreeSet};

use tally::{count, is_rate_bearing, is_user_denial, latency, sum_over, tokens_basis};
// Re-exported so this module's surface is `observe::extract::*`, exactly as `extract.py` is one
// module: the split into two files is the 400-line file budget and nothing else. `unused_imports` is
// allowed because the non-test build of this phase has no caller for the tallies `attribute` and
// `envelope` will consume in Phase 4; the tests below use every one of them today.
#[allow(unused_imports)]
pub(crate) use tally::{
    child_tokens_total, dedupe_usages, repeated_tool_calls, sum_tokens, sum_tokens_by_model,
    tool_class, tool_class_shares, total_tokens, EDIT_TOOLS, READ_TOOLS, SHELL_TOOLS,
};

use super::taskcat;
use super::types::{
    utc_day, InvocationObs, RunFacts, SessionFacts, ToolCall, ASSET_AGENT, ASSET_MCP_SERVER,
    ASSET_SKILL, FAILURE_CLASSES,
};

/// The main transcript was still being written when it was read (`extract.py:38`).
pub(crate) const OUTCOME_TRUNCATED: &str = "truncated";
/// The session compacted its context and did not then end its turn.
pub(crate) const OUTCOME_COMPACTED: &str = "compacted";
/// A tool call was cut off, or never resolved at all.
pub(crate) const OUTCOME_INTERRUPTED: &str = "interrupted";
/// The last assistant message stopped with `end_turn`.
pub(crate) const OUTCOME_COMPLETED: &str = "completed";
/// None of the above — the transcript does not say how the run ended.
pub(crate) const OUTCOME_UNKNOWN: &str = "unknown";

/// `sources/base.py:22`. Indexed out of [`FAILURE_CLASSES`] so the two cannot drift apart.
pub(super) const FAILURE_USER_DENIED: &str = FAILURE_CLASSES[2];
/// `sources/base.py:23`.
const FAILURE_INTERRUPTED: &str = FAILURE_CLASSES[3];
/// `sources/base.py:24`.
const FAILURE_UNKNOWN: &str = FAILURE_CLASSES[4];

/// The published tool-mix classes (`extract.py:43`, D2). Every one is present in
/// [`tool_class_shares`] even at zero, so the shape taskcat sees is stable.
pub(crate) const TOOL_CLASSES: [&str; 5] = ["edit", "read", "shell", "mcp", "other"];

/// A `(name, input_fingerprint)` group of this size or larger is a repeat (`extract.py:47`).
pub(crate) const REPEAT_THRESHOLD: u64 = 3;

/// Harness permission-mode spellings mapped to the gate's enum (`extract.py:49-56`).
const PERMISSION_MODES: [(&str, &str); 6] = [
    ("acceptEdits", "accept_edits"),
    ("bypassPermissions", "bypass"),
    ("dontAsk", "dont_ask"),
    ("plan", "plan"),
    ("default", "default"),
    ("auto", "auto"),
];

/// The gate's `enums.permission_mode` (`extract.py:255`). A source that pre-maps (the Codex
/// approval policy) passes through unchanged.
pub(crate) const PERMISSION_ENUM: [&str; 7] = [
    "default",
    "plan",
    "accept_edits",
    "bypass",
    "auto",
    "dont_ask",
    "unknown",
];

/// The gate's `enums.effort` (`extract.py:57`), minus `unknown` which is the fallback.
const EFFORTS: [&str; 5] = ["minimal", "low", "medium", "high", "xhigh"];

/// Derive [`RunFacts`] for the run rooted at `facts`.
///
/// `now_ms` is only the fallback for a session that carries no harness timestamp at all
/// (`extract.py:71-73`).
pub(crate) fn extract(facts: &SessionFacts, now_ms: i64) -> RunFacts {
    let tree = walk(facts);
    let calls: Vec<&ToolCall> = tree.iter().flat_map(|f| f.tool_calls.iter()).collect();
    let usages = dedupe_usages(&tree);
    let (first, last) = span(&tree, now_ms);
    RunFacts {
        session_key: facts.ref_.session_key.clone(),
        harness: facts.ref_.harness.clone(),
        harness_version: facts.harness_version.clone(),
        entrypoint_class: entrypoint_class(&facts.entrypoint).to_string(),
        effort: effort_class(&facts.effort).to_string(),
        permission_mode: permission_mode(&facts.permission_mode).to_string(),
        model: taskcat::allowlist_model(Some(dominant_model(&tree))).to_string(),
        observed_day: utc_day(first),
        first_ts_ms: first,
        last_ts_ms: last,
        run_outcome: run_outcome(facts).to_string(),
        turns: facts.user_turns,
        tool_calls: calls.len() as u64,
        tool_failures: count(&calls, is_rate_bearing),
        user_denials: count(&calls, is_user_denial),
        subagent_runs: tree.len() as u64 - 1,
        compactions: sum_over(&tree, |f| f.compactions),
        unpaired_tool_uses: count(&calls, |c| !c.paired()),
        repeated_tool_calls: repeated_tool_calls(&calls),
        tokens: sum_tokens(&usages),
        tokens_basis: tokens_basis(&usages).to_string(),
        tokens_by_model: sum_tokens_by_model(&usages),
        mcp_corroborations: merge_mcp_corroborations(&tree),
        tool_class_shares: tool_class_shares(&calls),
        invocations: invocations(facts),
        loaded_events: facts.loaded_events.clone(),
        in_band_assets: tree.iter().flat_map(|f| f.in_band_assets.clone()).collect(),
        lines_seen: sum_over(&tree, |f| f.lines_seen),
        lines_unknown_type: sum_over(&tree, |f| f.lines_unknown_type),
        bytes_read: sum_over(&tree, |f| f.bytes_read),
        parse_errors: sum_over(&tree, |f| f.parse_errors),
        truncated: facts.truncated,
        forbids: merge_forbids(&tree),
    }
}

// -- tree helpers --------------------------------------------------------------------------------

/// Depth-first, parent before children, in the order the source linked them (`extract.py:120-124`).
///
/// The order is part of the contract: [`dedupe_usages`] keeps the first entry on a tie, so
/// "parent first" is what decides which of two equal usages survives.
pub(crate) fn walk(facts: &SessionFacts) -> Vec<&SessionFacts> {
    let mut out = vec![facts];
    for child in &facts.children {
        out.extend(walk(child));
    }
    out
}

/// Harness-native MCP attribution markers, summed over the tree (`extract.py:140-145`). The keys
/// are server names and are **local only**.
pub(crate) fn merge_mcp_corroborations(tree: &[&SessionFacts]) -> BTreeMap<String, u64> {
    let mut out: BTreeMap<String, u64> = BTreeMap::new();
    for facts in tree {
        for (server, n) in &facts.mcp_attribution_counts {
            let slot = out.entry(server.clone()).or_insert(0);
            *slot = slot.saturating_add(*n);
        }
    }
    out
}

/// Union of every bucket of local-only names in the tree (`extract.py:148-153`), so the gate
/// checker sees a child's ids as well as the parent's.
pub(crate) fn merge_forbids(tree: &[&SessionFacts]) -> BTreeMap<String, BTreeSet<String>> {
    let mut out: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    for facts in tree {
        for (bucket, values) in &facts.forbids {
            out.entry(bucket.clone())
                .or_default()
                .extend(values.iter().cloned());
        }
    }
    out
}

/// `(first, last)` over the tree, falling back to `now_ms` (`extract.py:156-161`).
///
/// `last` is clamped up to `first` so a tree whose only stamps are out of order can never report a
/// negative duration.
fn span(tree: &[&SessionFacts], now_ms: i64) -> (i64, i64) {
    let first = tree.iter().filter_map(|f| f.first_ts_ms).min();
    let last = tree.iter().filter_map(|f| f.last_ts_ms).max();
    let first = first.unwrap_or(now_ms);
    (first, last.unwrap_or(first).max(first))
}

/// Most frequent model by response count in the MAIN transcript (`extract.py:164-178`).
///
/// The whole tree is used only when the main transcript carried no model. Sub-agents may run on a
/// different model, which is why the envelope also carries `tokens_by_model`. Ties break on the
/// smaller name so the result is deterministic. `"unknown"` when no response carried a model, which
/// [`taskcat::allowlist_model`] then reports as `"other"`.
pub(crate) fn dominant_model<'a>(tree: &[&'a SessionFacts]) -> &'a str {
    let main_has_models = tree.first().is_some_and(|f| !f.models.is_empty());
    let sources = if main_has_models { &tree[..1] } else { tree };
    let mut counts: BTreeMap<&str, u64> = BTreeMap::new();
    for facts in sources {
        for (model, n) in &facts.models {
            let slot = counts.entry(model.as_str()).or_insert(0);
            *slot = slot.saturating_add(*n);
        }
    }
    // BTreeMap iteration is by name, so `max_by_key` on the count alone would keep the LAST maximum;
    // the Python sorts by `(-count, name)` and takes the first, i.e. the smallest name among ties.
    counts
        .into_iter()
        .fold(None::<(&str, u64)>, |best, (model, n)| match best {
            Some((_, top)) if top >= n => best,
            _ => Some((model, n)),
        })
        .map(|(model, _)| model)
        .unwrap_or("unknown")
}

// -- run shape -----------------------------------------------------------------------------------

/// Entrypoint string -> closed enum, by substring, case-insensitively (`extract.py:252-262`).
///
/// Precedence is remote > ide > sdk > cli > unknown, so `"sdk-cli"` is `sdk` and `"remote-cli"` is
/// `remote`. The Python accepts `Optional[str]` and folds `None` into `""`; both are `unknown`, so
/// this takes `&str` and the empty string stands in for the missing value.
pub(crate) fn entrypoint_class(raw: &str) -> &'static str {
    let raw = raw.to_lowercase();
    if raw.contains("remote") {
        "remote"
    } else if raw.contains("vscode") || raw.contains("jetbrains") || raw.contains("ide") {
        "ide"
    } else if raw.contains("sdk") {
        "sdk"
    } else if raw.contains("cli") {
        "cli"
    } else {
        "unknown"
    }
}

/// Permission mode -> closed enum (`extract.py:268-273`).
///
/// Claude Code's camelCase values are mapped; a value already in [`PERMISSION_ENUM`] (a source that
/// pre-maps, like the Codex approval policy) passes through unchanged. Matching is exact:
/// `acceptedits` is `unknown`, not `accept_edits`.
pub(crate) fn permission_mode(raw: &str) -> &'static str {
    if let Some(known) = PERMISSION_ENUM.iter().find(|mode| **mode == raw) {
        return known;
    }
    PERMISSION_MODES
        .iter()
        .find(|(harness, _)| *harness == raw)
        .map(|(_, mapped)| *mapped)
        .unwrap_or("unknown")
}

/// Effort -> closed enum (`extract.py:276-279`).
///
/// Anything the gate does not list — including harness values it has never heard of — is `unknown`.
/// The contract is silent on effort; this is what keeps the payload gate-clean regardless.
pub(crate) fn effort_class(raw: &str) -> &'static str {
    EFFORTS
        .iter()
        .find(|effort| **effort == raw)
        .copied()
        .unwrap_or("unknown")
}

/// Decision table, first match wins (`extract.py:286-297`):
/// truncated > compacted > interrupted > completed > unknown.
///
/// Only the transcript `facts` itself is consulted — children have their own outcome, which is what
/// [`child_failure_class`] reads.
pub(crate) fn run_outcome(facts: &SessionFacts) -> &'static str {
    if facts.truncated {
        return OUTCOME_TRUNCATED;
    }
    if facts.compactions > 0 && facts.last_stop_reason.as_deref() != Some("end_turn") {
        return OUTCOME_COMPACTED;
    }
    if interrupted_at_end(&facts.tool_calls) {
        return OUTCOME_INTERRUPTED;
    }
    if facts.last_stop_reason.as_deref() == Some("end_turn") {
        return OUTCOME_COMPLETED;
    }
    OUTCOME_UNKNOWN
}

/// Any call without a result, or the last call in transcript order marked interrupted
/// (`extract.py:300-304`).
///
/// An interrupt that is not the last call does not make a finished run interrupted: the run went on
/// afterwards.
fn interrupted_at_end(calls: &[ToolCall]) -> bool {
    calls.iter().any(|c| !c.paired()) || calls.last().is_some_and(|c| c.interrupted)
}

// -- invocations ---------------------------------------------------------------------------------

/// Explicit asset invocations over the tree: parent's calls first, then each child's
/// (`extract.py:310-323`).
///
/// The precedence within one call is skill > server > agent_type, and a call that names none of the
/// three is not an invocation of anything. Names are **local only**; `attribute` hashes them.
///
/// The three tests are on a *non-empty* name because the Python's `if call.skill:` is a truthiness
/// test, not a presence test. The difference is not cosmetic: with a presence test, a call carrying
/// `skill = Some("")` alongside a real `server` is attributed to the wrong asset *type*, and the
/// empty name then seeds a phantom `AssetKey` whose `name_hash` reaches `assets[]` — neither
/// `seed_invocations` here nor `attribute.py` guards it, because upstream never emits one.
pub(crate) fn invocations(facts: &SessionFacts) -> Vec<InvocationObs> {
    let mut out = Vec::new();
    for call in &facts.tool_calls {
        if let Some(skill) = call.skill.as_deref().filter(|name| !name.is_empty()) {
            out.push(simple_invocation(ASSET_SKILL, skill, call));
        } else if let Some(server) = call.server.as_deref().filter(|name| !name.is_empty()) {
            out.push(simple_invocation(ASSET_MCP_SERVER, server, call));
        } else if call
            .agent_type
            .as_deref()
            .is_some_and(|name| !name.is_empty())
        {
            out.push(agent_invocation(call, linked_child(facts, call)));
        }
    }
    for child in &facts.children {
        out.extend(invocations(child));
    }
    out
}

fn simple_invocation(asset_type: &str, name: &str, call: &ToolCall) -> InvocationObs {
    InvocationObs {
        asset_type: asset_type.to_string(),
        name: name.to_string(),
        ts_ms: call.ts_ms,
        latency_ms: latency(call),
        failure_class: call.failure_class.clone(),
        is_async: call.is_async,
        ..Default::default()
    }
}

/// The direct child this `Agent` call spawned, by `child_key` or by the child's own `toolUseId`
/// (`extract.py:326-332`).
///
/// Both keys are tried per child, in child order, so the first child that matches either wins.
fn linked_child<'a>(facts: &'a SessionFacts, call: &ToolCall) -> Option<&'a SessionFacts> {
    facts.children.iter().find(|child| {
        let by_key = call
            .child_key
            .as_deref()
            .filter(|key| !key.is_empty())
            .is_some_and(|key| child.ref_.session_key == key);
        let by_meta = child.ref_.child_meta.get("toolUseId") == Some(&call.tool_use_id);
        by_key || by_meta
    })
}

/// The agent invocation for a spawn (`extract.py:335-348`).
///
/// The parent's result is only a spawn ack (D4): outcome and tokens come from the child when one is
/// linked. The parent's own failure class (a denied or failed spawn) still takes precedence, because
/// then there is no child run to speak of.
fn agent_invocation(call: &ToolCall, child: Option<&SessionFacts>) -> InvocationObs {
    let mut obs = simple_invocation(ASSET_AGENT, call.agent_type.as_deref().unwrap_or(""), call);
    if let Some(child) = child {
        // `failure or _child_failure_class(child)` is a truthiness `or`, so an empty parent class
        // falls through to the child's. Keeping `Some("")` would put a value outside FAILURE_CLASSES
        // on the wire, which the gate rejects.
        obs.failure_class = obs
            .failure_class
            .filter(|class| !class.is_empty())
            .or_else(|| child_failure_class(child).map(str::to_string));
        obs.corroborated = child
            .ref_
            .child_meta
            .get("corroborated")
            .is_some_and(|value| value.to_lowercase() == "true");
        obs.child_tokens_total = child_tokens_total(child);
    }
    obs
}

/// Child outcome -> failure class (`extract.py:351-358`).
///
/// Never a rate-bearing class: a child's own tool errors are counted on the tools it called, not on
/// the agent as a whole, so a sub-agent cannot inflate its own type's non-success rate twice.
fn child_failure_class(child: &SessionFacts) -> Option<&'static str> {
    match run_outcome(child) {
        OUTCOME_INTERRUPTED => Some(FAILURE_INTERRUPTED),
        OUTCOME_TRUNCATED | OUTCOME_UNKNOWN => Some(FAILURE_UNKNOWN),
        _ => None,
    }
}

#[cfg(test)]
#[path = "extract_tests.rs"]
mod tests;
